File: TinkoffChat/ThemesViewController.m
//
//  ThemesViewController.m
//  TinkoffChat
//
//  Created by me on 14/10/2018.
//  Copyright © 2018 gladkikh. All rights reserved.
//

#import "ThemesViewController.h"

@interface ThemesViewController ()

@end

@implementation ThemesViewController

@synthesize delegate;
@synthesize model; // = _model

- (IBAction)buttonTheme1Touched:(id)sender {
    self.view.backgroundColor = self.model.theme1;
    [self.delegate themesViewController:self didSelectTheme:self.model.theme1];
}
- (IBAction)buttonTheme2Touched:(id)sender {
    self.view.backgroundColor = self.model.theme2;
    [self.delegate themesViewController:self didSelectTheme:self.model.theme2];
}
- (IBAction)buttonTheme3Touched:(id)sender {
    self.view.backgroundColor = self.model.theme3;
    [self.delegate themesViewController:self didSelectTheme:self.model.theme3];
}
- (IBAction)closeButtonTapped:(id)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.model = [[Themes alloc] init];
    self.model.theme1 = [UIColor colorWithRed:255.0f/255.0f green:0.0f/255.0f blue:0.0f/255.0f alpha:1.0f];
    self.model.theme2 = [UIColor colorWithRed:0.0f/255.0f green:255.0f/255.0f blue:0.0f/255.0f alpha:1.0f];
    self.model.theme3 = [UIColor colorWithRed:0.0f/255.0f green:0.0f/255.0f blue:255.0f/255.0f alpha:1.0f];
}

//-(void)dealloc {
//    self.model = nil;
//    [super dealloc];
//}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: TinkoffChat/ThemesViewController.swift
//
//  ThemesViewController.swift
//  TinkoffChatProductTwo
//
//  Created by me on 14/10/2018.
//  Copyright © 2018 gladkikh. All rights reserved.
//

import UIKit

class ThemesViewController: UIViewController {
    
    var model: Themes?
    var themeChanged: ((UIColor?) -> Void)?
    
    func applyColorChanges(theme: UIColor?) {
        if let color = theme {
            view.backgroundColor = color
            let queue = DispatchQueue.global(qos: .utility)
            queue.async{
                do {
                    try UserDefaults.standard.setValue(NSKeyedArchiver.archivedData(withRootObject: color, requiringSecureCoding: false), forKey: "ApplicationTheme")
                    UserDefaults.standard.synchronize()
                } catch {
                    print("Cannot save application theme to UserDefaults")
                }
            }
            AppDelegate.changeApplicationColor(color: color)
        }
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        model = Themes()
        model?.theme1 = UIColor(red: 255.0/255.0, green: 0.0/255.0, blue: 0.0/255.0, alpha: 1.0)
        model?.theme2 = UIColor(red: 0.0/255.0, green: 255.0/255.0, blue: 0.0/255.0, alpha: 1.0)
        model?.theme3 = UIColor(red: 0.0/255.0, green: 0.0/255.0, blue: 255.0/255.0, alpha: 1.0)
    }
    @IBAction func buttonTheme1Touched(_ sender: Any) {
        applyColorChanges(theme: model?.theme1)
        themeChanged?(model?.theme1)
    }
    @IBAction func buttonTheme2Touched(_ sender: Any) {
        applyColorChanges(theme: model?.theme2)
        themeChanged?(model?.theme2)
    }
    @IBAction func buttonTheme3Touched(_ sender: Any) {
        applyColorChanges(theme: model?.theme3)
        themeChanged?(model?.theme3)
    }
    @IBAction func closeButtonTapped(_ sender: Any) {
        self.dismiss(animated: true, completion: nil)
    }

}
